src/ol/style/expressions.js

```diff
@@ -744,17 +770,16 @@ Operators['interpolate'] = {
     assertUniqueInferredType(args, outputType);
 
     const input = expressionToGlsl(context, args[1]);
-    let result = null;
+    const exponent = numberToGlsl(interpolation);
+
+    let result = '';
     for (let i = 2; i < args.length - 2; i += 2) {
       const stop1 = expressionToGlsl(context, args[i]);
-      const output1 = expressionToGlsl(context, args[i + 1], outputType);
+      const output1 =
+        result || expressionToGlsl(context, args[i + 1], outputType);
       const stop2 = expressionToGlsl(context, args[i + 2]);
       const output2 = expressionToGlsl(context, args[i + 3], outputType);
-      result = `mix(${
-        result || output1
-      }, ${output2}, pow(clamp((${input} - ${stop1}) / (${stop2} - ${stop1}), 0.0, 1.0), ${numberToGlsl(
-        interpolation
-      )}))`;
+      result = `mix(${output1}, ${output2}, pow(clamp((${input} - ${stop1}) / (${stop2} - ${stop1}), 0.0, 1.0), ${exponent}))`;
```

Comment:
Ideally we would not have to repeat the `input` expression for each stop.  For example, in the `cog-math.js` example, the interpolate expression serializes to this:

```
mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(mix(vec4(0.7490196078431373, 0.7490196078431373, 0.7490196078431373, 1.0), vec4(0.8588235294117647, 0.8588235294117647, 0.8588235294117647, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - -0.2) / (-0.1 - -0.2), 0.0, 1.0), 1.0)), vec4(1.0, 1.0, 0.8784313725490196, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - -0.1) / (0.0 - -0.1), 0.0, 1.0), 1.0)), vec4(1.0, 0.9803921568627451, 0.8, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.0) / (0.025 - 0.0), 0.0, 1.0), 1.0)), vec4(0.9294117647058824, 0.9098039215686274, 0.7098039215686275, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.025) / (0.05 - 0.025), 0.0, 1.0), 1.0)), vec4(0.8705882352941177, 0.8509803921568627, 0.611764705882353, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.05) / (0.075 - 0.05), 0.0, 1.0), 1.0)), vec4(0.8, 0.7803921568627451, 0.5098039215686274, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.075) / (0.1 - 0.075), 0.0, 1.0), 1.0)), vec4(0.7411764705882353, 0.7215686274509804, 0.4196078431372549, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.1) / (0.125 - 0.1), 0.0, 1.0), 1.0)), vec4(0.6901960784313725, 0.7607843137254902, 0.3803921568627451, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.125) / (0.15 - 0.125), 0.0, 1.0), 1.0)), vec4(0.6392156862745098, 0.8, 0.34901960784313724, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.15) / (0.175 - 0.15), 0.0, 1.0), 1.0)), vec4(0.5686274509803921, 0.7490196078431373, 0.3215686274509804, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.175) / (0.2 - 0.175), 0.0, 1.0), 1.0)), vec4(0.5019607843137255, 0.7019607843137254, 0.2784313725490196, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.2) / (0.25 - 0.2), 0.0, 1.0), 1.0)), vec4(0.4392156862745098, 0.6392156862745098, 0.25098039215686274, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.25) / (0.3 - 0.25), 0.0, 1.0), 1.0)), vec4(0.3803921568627451, 0.5882352941176471, 0.21176470588235294, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.3) / (0.35 - 0.3), 0.0, 1.0), 1.0)), vec4(0.30980392156862746, 0.5411764705882353, 0.1803921568627451, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.35) / (0.4 - 0.35), 0.0, 1.0), 1.0)), vec4(0.25098039215686274, 0.49019607843137253, 0.1411764705882353, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.4) / (0.45 - 0.4), 0.0, 1.0), 1.0)), vec4(0.18823529411764706, 0.43137254901960786, 0.10980392156862745, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.45) / (0.5 - 0.45), 0.0, 1.0), 1.0)), vec4(0.12941176470588237, 0.3803921568627451, 0.07058823529411765, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.5) / (0.55 - 0.5), 0.0, 1.0), 1.0)), vec4(0.058823529411764705, 0.32941176470588235, 0.0392156862745098, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.55) / (0.6 - 0.55), 0.0, 1.0), 1.0)), vec4(0.0, 0.27058823529411763, 0.0, 1.0), pow(clamp((((color.g - color.r) / (color.g + color.r)) - 0.6) / (0.65 - 0.6), 0.0, 1.0), 1.0))
```

We could factor out this:
```
float interpolateInput = (color.g - color.r) / (color.g + color.r);
```

And use `interpolateInput` in the expression instead.  We could make this work by having the `toGlsl` function append additional statements to the context that would be included when composing the shader.  Or we could have `toGlsl` return more than just a string expression.

I'm not sure how to measure whether having the fragment shader evaluate this `(color.g - color.r) / (color.g + color.r)` 20 extra times per pixel has a significant impact.  Curious if others think this type of thing is worth optimizing.

Reply:
This was also something I was wondering about. AFAIK the GPUs do some optimizations when compiling shaders but this is hardware dependant I think? Also some operations are more costly than others. Here though, I don't know if the extra computations have any impact at all. Probably @IvanSanchez would know more about this?